### DISCUSSION TITLE
`obj.has()` or `arr.upsert()`

### DISCUSSION THREAD

**Discussion by macmillen** [streamich/json-joy#691] [Ideas]
I often find myself having a field that is optional and wanting to set data on that field. But to do that I first need to check if that field exists and for that I currently convert the node to a `view()` and then I check if it exists before initializing the field. See this example:
``` ts
if (!("children" in node.view())) node.set({ children: [] });
node.arr("children").ins(0, blocks);
```

It would be great if there was a `has()` operation on the `ObjApi`. Another idea could be maybe to have something like an `upsert()` operation:

Option 1:
``` ts
if (!(node.has("children")) node.set({ children: [] });
``` 
Option 2:
``` ts
node.arr("children").upsert(0, blocks)
``` 

Option 2 would already fail when trying to access the prop with `.arr("children")` so it would need to be a different API.

**Comment by streamich** [comment]
`ObjApi.has()` just landed in `17.45.0`.

**Comment by streamich** [comment]
By "upsert" you mean appending to the end of the array? In that case, yes, I think we should implement `ArrApi.push()`. (I think `VecApi` already has it but `ArrApi` is missing it.)

**Reply by macmillen** [reply]
With upsert I mean:
1. If a node already exists at the path update the content
2. If the node doesn't exist yet create a new node
